File: app/src/main/java/closer/vlllage/com/closer/handler/group/PhysicalGroupHandler.java
package closer.vlllage.com.closer.handler.group;

import com.google.android.gms.maps.model.LatLng;

import closer.vlllage.com.closer.handler.bubble.BubbleType;
import closer.vlllage.com.closer.handler.bubble.MapBubble;
import closer.vlllage.com.closer.handler.data.SyncHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Group;

public class PhysicalGroupHandler extends PoolMember {

    public void createPhysicalGroup(LatLng latLng) {
        Group group = $(StoreHandler.class).create(Group.class);
        group.setName("");
        group.setAbout("");
        group.setPublic(true);
        group.setPhysical(true);
        group.setLatitude(latLng.latitude);
        group.setLongitude(latLng.longitude);
        $(StoreHandler.class).getStore().box(Group.class).put(group);
        $(SyncHandler.class).sync(group, this::openGroup);
    }

    private void openGroup(String groupId) {
        $(GroupActivityTransitionHandler.class).showGroupMessages(null, groupId, true);
    }

    public MapBubble physicalGroupBubbleFrom(Group group) {
        if (group.getLatitude() == null || group.getLongitude() == null) {
            return null;
        }
        MapBubble mapBubble = new MapBubble(new LatLng(group.getLatitude(), group.getLongitude()), "Group", "");
        mapBubble.setType(BubbleType.PHYSICAL_GROUP);
        mapBubble.setPinned(true);
        mapBubble.setTag(group);
        return mapBubble;
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/PhysicalGroupHandler.kt
package closer.vlllage.com.closer.handler.group

import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.bubble.BubbleType
import closer.vlllage.com.closer.handler.bubble.MapBubble
import closer.vlllage.com.closer.handler.data.OnSyncResult
import closer.vlllage.com.closer.handler.data.SyncHandler
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.Group
import closer.vlllage.com.closer.store.models.GroupMessage
import closer.vlllage.com.closer.store.models.GroupMessage_
import com.google.android.gms.maps.model.LatLng
import com.queatz.on.On
import io.reactivex.Single

class PhysicalGroupHandler constructor(private val on: On) {

    fun createPhysicalGroup(latLng: LatLng, isPublic: Boolean = true, hub: Boolean = false, name: String = "", onSyncResult: OnSyncResult? = null) {
        val group = on<StoreHandler>().create(Group::class.java)
        group!!.name = name
        group.hub = hub
        group.about = ""
        group.isPublic = isPublic
        group.physical = true
        group.latitude = latLng.latitude
        group.longitude = latLng.longitude
        on<StoreHandler>().store.box(Group::class).put(group)
        on<SyncHandler>().sync(group, onSyncResult ?: { id -> openGroup(id) })
    }

    fun physicalGroupBubbleFrom(group: Group): MapBubble? {
        if (group.latitude == null || group.longitude == null) {
            return null
        }
        val mapBubble = MapBubble(LatLng(group.latitude!!, group.longitude!!), "Group", "")
        mapBubble.type = BubbleType.PHYSICAL_GROUP
        mapBubble.isPinned = true
        mapBubble.tag = group
        return mapBubble
    }

    fun physicalGroupName(group: Group): Single<String> = when {
        group.name.isNullOrBlank() -> on<StoreHandler>().store.box(GroupMessage::class).query(
                GroupMessage_.to.equal(group.id ?: "").and(GroupMessage_.text.notNull())
        )
                .order(GroupMessage_.created)
                .build().findFirst()?.text?.let { on<GroupMessageParseHandler>().parseString(it).map { "\"$it\"" } }
                    ?: Single.just(on<ResourcesHandler>().resources.getString(R.string.talk_here))
        else -> Single.just(group.name)
    }

    private fun openGroup(groupId: String) {
        on<GroupActivityTransitionHandler>().showGroupMessages(null, groupId, true)
    }
}
